File: mazes/maze-01/ruby/maze.rb
#!/usr/bin/ruby

require 'optparse'

DEFAULT_WIDTH = 10
DEFAULT_HEIGHT = 10
DEFAULT_SEED = rand(0xFFFF_FFFF)
DEFAULT_ANIMATE = false
DEFAULT_DELAY = 0.02

# ===================================================================
# Class Maze defines basic behavior to which a maze should conform.
# It provides basic initialization/construction for the maze class, 
# and provides a method for drawing ASCII mazes. 
#
# Specific "maze-carving" techniques are implemented in subclasses.
# ====================================================================
class Maze
	# +++++++++++++++++++++++++++
	# Configure class variables.
	# +++++++++++++++++++++++++++
	@@N, @@S, @@E, @@W = 1, 2, 4, 8 
	@@DX = { @@E => 1, @@W => -1, @@N => 0, @@S => 0 }
	@@DY = { @@E => 0, @@W => 0, @@N => -1, @@S => 1 }
	@@OPPOSITE = { @@E => @@W, @@W => @@E, @@N => @@S, @@S => @@N }

	def self.N; @@N; end
	def self.S; @@S; end
	def self.E; @@E; end
	def self.W; @@W; end
	def self.DX; @@DX; end
	def self.DY; @@DY; end
	def self.OPPOSITE; @@OPPOSITE; end
	
	# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	# Initialize a new 2D maze with the given width and height.
	#
	# Default seed value will give "random" behavior.
	# User-supplied seed value will give "deterministic behavior.
	# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	def initialize( w=DEFAULT_WIDTH, h=DEFAULT_HEIGHT, s=DEFAULT_SEED )
		@width = w
		@height = h
		@seed = s

		srand(@seed)	

		@grid = Array.new(h) { Array.new(w,0) }
	end
	
	attr_reader :width, :height, :seed

	# +++++++++++++++++++++++++++++++++++++++++++++++++++++++
	# Draw the grid, starting in the upper-left hand corner.
	# +++++++++++++++++++++++++++++++++++++++++++++++++++++++
	def draw
		#
		# Draw the "top" line.
		#
		puts " " + "_" * (@width * 2 - 1)
	
		#	
		# Draw each of the rows.
		#
		@height.times do |y|
			print "|"
			@width.times do |x|
				# render "bottom" using "S" switch
				print( (@grid[y][x] & @@S != 0) ? " " : "_" )
			
				# render "side" using "E" switch	
				if @grid[y][x] & @@E != 0
					print( ( (@grid[y][x] | @grid[y][x+1]) & @@S != 0 ) ? " " : "_" )
				else
					print "|"
				end
			end
			puts
		end

		# 
		# Output metadata
		#
		puts metadata
	end

	# ++++++++++++++++++++++++++++++ 
	# Output generic maze metadata.
	# ++++++++++++++++++++++++++++++ 
	def metadata
	    "#{$0} #{@width} #{@height} #{@seed} #{@delay}"
	end
end

# =========================================================================
# Class BackTracker implements a simple recursive back-tracking algorithm
# to draw ASCII mazes. The algorithms works as a "depth-first" search
# of the "tree" or "graph" representing the maze.
# 
# A possible optimization might be to implement a "breadth-first" search.
# =========================================================================
class BackTracker < Maze

	# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	# Initialize a new 2D maze with the given width and height.
	#
	# Default seed value will give "random" behavior.
	# User-supplied seed value will give "deterministic behavior.
	# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	def initialize( w=DEFAULT_WIDTH, h=DEFAULT_HEIGHT, s=DEFAULT_SEED, a=DEFAULT_ANIMATE, d=DEFAULT_DELAY )
		#
		# Invoke super-constructor
		#
		super(w,h,s)

		# 
		# Only prepare the maze beforehand if we are doing "static" (i.e., animate=false) drawing
		#
		@delay = d
		@animate = a
		if not @animate
		      carve_passage_from(0,0)
		end
	end

	# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	# Recursively carve passages through the maze, starting at (x,y).
	#
	# Algorithm halts when all "cells" in the maze have been visited.
	# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	def carve_passage_from(x,y)
		directions = [@@N,@@S,@@E,@@W].sort_by{rand}
		directions.each do |direction|
			#
			# Render updates of the maze on a "cell-by-cell" basis
			#
			if @animate
			      display(x,y)
			      sleep @delay
			end

			dx,dy = x + @@DX[direction], y + @@DY[direction]
			if dy.between?(0,@grid.length-1) && dx.between?(0,@grid[dy].length-1) && @grid[dy][dx] == 0
				@grid[y][x] |= direction			# "open" the wall from current cell;
				@grid[dy][dx] |= @@OPPOSITE[direction]		# "open" the wall from adjacent cell;
				carve_passage_from(dx,dy)
			end
		end

		# 
		# Make one final call to "update" to display last cell.
		# Set the coords to (-1,-1) so the cell is left "blank" with no cursor.
		#
		if @animate
		      display(-1,-1)
		end
	end

	# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	# Display needs the (x,y) coordinates of where it is presently rendering, in 
	# order to color the "current cursor" cell a different color (in this case, 
	# red). We've already used the symbols "x" and "y" in a previous implementation
	# of this algorithm, so we'll name them "i" and "j" in the method signature instead.
	# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	def display(i,j)
 	      print "\e[H"
	      puts " " + "_" * (@width * 2 - 1)
	      
	      @grid.each_with_index do |row,y|
	            print "|"
		    row.each_with_index do |cell,x|
		          # 
			  # Color gray if empty, red if "current" cursor
			  #
			  if cell == 0
			        print "\e[47m" 
			  elsif x == i and y == j
 			        print "\e[41m"
			  end

			  # render "bottom" using "S" switch
			  print( (@grid[y][x] & @@S != 0) ? " " : "_" )
			  
			  # render "side" using "E" switch
			  if @grid[y][x] & @@E != 0
			        print( ( (@grid[y][x] | @grid[y][x+1]) & @@S != 0 ) ? " " : "_" )
			  else
			        print "|"
			  end

			  # 
			  # Stop coloring
			  #
			  if cell == 0 or ( x == i and y == j )
			        print "\e[m" 
			  end
		    end
		    puts
	      end
	end

	# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ 
	# Method only needs to be overridden if we are animating.
	# 
	# If we are drawing the maze statically, defer to the superclass.
	# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ 
	def draw

	    # 
	    # Clear the screen.
	    #
	    print "\e[2J"
	    if not @animate
	       
	        # 
		# Move to upper left and defer to superclass
		#		
	        print "\e[H"
		super()
	    else
		#
		# If we are animating, clear the screen and start carving!
		#
		carve_passage_from(0,0)

		#
		# Output maze metadata
		#
		puts metadata
	    end
	end

	# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ 
	# Override metadata to inform what type of maze we are carving
	# +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	def metadata
	    "#{$0} #{@width} #{@height} #{@seed} #{@delay} [BackTracker]"
	end

	protected :carve_passage_from
end

# ============================
# Command line code goes here
# ============================
OPTIONS  = {
	:w => DEFAULT_WIDTH,
	:h => DEFAULT_HEIGHT,
	:s => DEFAULT_SEED,
	:a => DEFAULT_ANIMATE,
	:d => DEFAULT_DELAY
}

if __FILE__ == $0
	ARGV.options do |o|
		# parse the command line options
		o.separator ""
		o.on("-w","--width=[value]", Integer, "Width of maze (default: " + DEFAULT_WIDTH.to_s + ")") 		{ |OPTIONS[:w]| }
		o.on("-h","--height=[value]", Integer, "Height of maze (default: " + DEFAULT_HEIGHT.to_s+ ")")		{ |OPTIONS[:h]| }
		o.on("-s","--seed=[value]", Integer, "User-defined seed will model deterministic behavior (default: " + DEFAULT_SEED.to_s + ")")	{ |OPTIONS[:s]| }
		o.on("-a","--[no-]animated", true.class, "Animate rendering (default: " + DEFAULT_ANIMATE.to_s + ")") 	{ |OPTIONS[:a]| }
		o.on("-d","--delay=[value]", Float, "Animation delay (default: " + DEFAULT_DELAY.to_s + ")")   	 	{ |OPTIONS[:d]| }
		o.separator ""
		o.parse!

		# sanitize the input from the command line
		good = true
		if OPTIONS[:w] == "" or OPTIONS[:w] == nil 
			good = false
		elsif OPTIONS[:h] == "" or OPTIONS[:h] == nil 
			good = false
		elsif OPTIONS[:s] == "" or OPTIONS[:s] == nil
			good = false
		elsif OPTIONS[:d] == "" or OPTIONS[:d] == nil
		        good = false
  		end

		if good
			# build and draw a new back-tracking maze
			BackTracker.new(w=OPTIONS[:w], h=OPTIONS[:h], s=OPTIONS[:s], a=OPTIONS[:a], d=OPTIONS[:d]).draw
		else
			puts o
		end
	end
end

File: mazes/maze-01/python/maze.py
#!/usr/bin/python

import random
import sys
import time
from optparse import OptionParser

DEFAULT_WIDTH = 10
DEFAULT_HEIGHT = 10 
DEFAULT_SEED = None   # using None will seed PRNG with current system time
DEFAULT_ANIMATE = False
DEFAULT_DELAY = 0.02

class Maze(object):
    """Class Maze defines basic behavior to which a maze should conform.
It provides basic initialization/construction for the maze class, 
and provides a method for drawing ASCII mazes.

Specific "maze-carving" techniques are implemented in subclasses."""

    #
    # Configure class variables
    #
    N,S,E,W = 1,2,4,8
    DX = { E:+1, W:-1, N:0, S:0 }
    DY = { E:0, W:0, N:-1, S:+1 }
    OPPOSITE = { E:W, W:E, N:S, S:N } 

    def __init__(self,w=DEFAULT_WIDTH,h=DEFAULT_HEIGHT,s=DEFAULT_SEED):
        """Initialize a new 2D maze with the given width and height.

Default seed will give 'random' behavior.
User-supplied seed will give deterministic behavior."""

        # initialize the instance variables
        self.width = DEFAULT_WIDTH if w is None else w
        self.height = DEFAULT_HEIGHT if h is None else h 
        self.seed = DEFAULT_SEED if s is None else s 

        # seed the PRNG
        random.seed(self.seed)

        # build the grid to hold the maze
        self.grid = [[0 for col in range(self.width)] for row in range(self.height)]

    def draw(self,):
        """Draw the grid, starting in the upper-left hand corner."""

        #
        # Draw the top line
        #
        buffer = []; out = " "
        for i in range(2*self.width - 1):
            out += "_" 
        buffer.append(out)

        #
        # Draw each of the rows.
        #
        for j in range(self.height):
            out = "|"
            for i in range(self.width):
                # draw the "bottom" using S switch
                out += " " if ((self.grid[j][i] & Maze.S) != 0) else "_"

                # draw the "side" using E switch
                if (self.grid[j][i] & Maze.E) != 0:
                    out += " " if (((self.grid[j][i] | self.grid[j][i+1]) & Maze.S) != 0) else "_"
                else:
                    out += "|"
     
            buffer.append(out)

        # 
        # Output generic maze metadata.
        #
        buffer.append(self.metadata())
        print "\r\n".join(buffer)

    def metadata(self,):
        """Generate maze metadata."""
        return  " ".join([sys.argv[0],str(self.width),str(self.height),str(self.seed)])

class BackTracker(Maze):
    """Class BackTracker implements a simple recursive back-tracking algorithm 
to draw ASCII mazes. The algorithm works as a "depth-first" search
of the "tree" or "graph" representing the maze.

A possible optimization might be to implement a "breadth-first" search.
"""

    def __init__(self,w=DEFAULT_WIDTH,h=DEFAULT_HEIGHT,s=DEFAULT_SEED,a=DEFAULT_ANIMATE,d=DEFAULT_DELAY):
        """Initialize a new 2D maze with the given width and height.

Default seed value will give "random" behavior.
User-supplied seed value will give "deterministic" behavior.
"""
        #
        # Invoke super-constructor.
        #
        super(BackTracker,self).__init__(w,h,s)

        #
        # Only prepare the maze beforehand if we are doing "static" (i.e., animate=false) drawing
        #
        self.delay = d
        self.animate = a
        if not self.animate:
            self.carve_passage_from(0,0)


    def metadata(self,):
        """Override metadata to inform what type of maze we are carving."""
        return Maze.metadata(self) + " [BackTracker]"

    def carve_passage_from(self,x,y):
        """Recursively carve passages through the maze, starting at (x,y).

Algorithm halts when all "cells" in the maze have been visited."""

        # 
        # Randomly shuffle the directions.
        #
        directions = [Maze.N, Maze.S, Maze.E, Maze.W]
        random.shuffle(directions)    

        #
        # Step through the maze, and recursively invoke the procedure.
        #
        for direction in directions:
            #
            # Render updates of the maze on a "cell-by-cell" basis
            if self.animate:
                self.display(x,y)
                time.sleep(self.delay)

            dx,dy = x + Maze.DX[direction], y + Maze.DY[direction]
            if dy >= 0 and dy < self.height and dx >= 0 and dx < self.width and self.grid[dy][dx] == 0:
                self.grid[y][x] |= direction
                self.grid[dy][dx] |= Maze.OPPOSITE[direction]
                self.carve_passage_from(dx,dy)

        #
        # Make one final call to "update" to display last cell.
        # Set the coords to (-1,-1) so the cell is left "blank" with no cursor.
        #
        if self.animate:
            self.display(-1,-1)

    def display(self,i,j):
        """Display needs the (x,y) coordinate of where it is presently rendering, in 
order to color the "current cursor" cell a different color (in this case, 
red). We've already used the symbols "x" and "y" in a previous implementation
of this algorithm, so we'll name them "i" and "j" in the method signature instead.
"""

        sys.stdout.write("\x1b[H")
        buffer = []; out = " "
        for c in range(2*self.width - 1):
            out += "_"
        buffer.append(out)

        #
        # Step through the cells of the maze
        #
        for y in range(self.height):
            out = "|"
            for x in range(self.width):
                # 
                # Color gray if empty, red if "current" cursor
                #
                if self.grid[y][x] == 0:
                    out += "\x1b[47m"
                elif x == i and y == j:
                    out += "\x1b[41m"

                # render "bottom" using "S" switch
                out += " " if ((self.grid[y][x] & Maze.S) != 0) else "_"

                # render "side" using "E" switch
                if ( self.grid[y][x] & Maze.E ) != 0:
                    out += " " if (((self.grid[y][x] | self.grid[y][x+1]) & Maze.S) != 0) else "_"
                else:
                    out += "|"

                #
                # Stop coloring
                #
                if self.grid[y][x] == 0 or ( x == i and y == j ):
                    out += "\x1b[m"

            buffer.append(out)

        buffer.append("")
        sys.stdout.write("\r\n".join(buffer))
   
    def draw(self,):
        """Method only needs to be overridden if we are animating.

If we are drawing the maze staticially, defer to the superclass.
"""
        #
        # Clear the screen.
        #
        sys.stdout.write("\x1b[2J")

        if not self.animate:
            #
            # Move to upper left and defer to superclass
            #
            sys.stdout.write("\x1b[H")
            Maze.draw(self)

        else:
            # 
            # If we are animating, clear the screen and start carving:
            #
            BackTracker.carve_passage_from(self,0,0)

            #
            # Output maze metadata
            #
            print self.metadata()

# 
# Parse the command line arguments
#
parser = OptionParser()
parser.set_conflict_handler("resolve")
parser.add_option("-w","--width",dest="width",help="width of maze",metavar="<maze-width>")
parser.add_option("-h","--height",dest="height",help="height of maze",metavar="<maze-height>") 
parser.add_option("-s","--seed",dest="seed",help="seed for PRNG generator",metavar="<prng-seed>")
parser.add_option("-a","--animate",action="store_true",dest="animate",help="animated rendering?",metavar="<animated>")
parser.add_option("-d","--delay",dest="delay",help="animation delay",metavar="<animation-delay>")
(options, args) = parser.parse_args()

#
# Set the arguments we will use
#
w = DEFAULT_WIDTH if options.width is None else int(options.width)
h = DEFAULT_HEIGHT if options.height is None else int(options.height)
s = DEFAULT_SEED if options.seed is None else int(options.seed)
a = True if options.animate else False
d = DEFAULT_DELAY if options.delay is None else float(options.delay)

#
# Build and draw a new maze
#
m = BackTracker(w,h,s,a,d)
m.draw()
